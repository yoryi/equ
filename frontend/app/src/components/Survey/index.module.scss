@import "src/styles/mixins/breakpoints";
@import "src/styles/mixins/icon-box";

:local(.container) {
  display: flex;
  flex-direction: column;
  align-items: center;
  flex-basis: 400px;
  max-width: 400px;
  font-family: Ilisarniq, sans-serif;
  text-align: center;
  padding: 0 32px;

  @include breakpoint('small') {
    padding: 28px 0;
  }

  @include breakpoint('small') {
    padding: 0;
  }

  :local(.iconBox) {
    @include icon-box;
    margin-bottom: 28px;
  }

  :local(.title) {
    font-weight: 600;
    font-size: 20px;
    line-height: 20px;
    color: var(--main-dark);
  }

  :local(.description) {
    font-size: 16px;
    line-height: 24px;
    color: var(--dark-grey);
    margin: 0;

    @include breakpoint('small') {
      font-size: 14px;
    }
  }

  :local(.button) {
    margin-top: 28px;

    @include breakpoint('small') {
      margin-top: 12px;
    }
  }
}
