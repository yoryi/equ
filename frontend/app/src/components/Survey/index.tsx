import * as mui from "@material-ui/core"
import * as React from "react"
import { useTranslation } from "react-i18next"
import { useSelector } from "react-redux"

import { ReactComponent as Book } from "../../assets/book.svg"
import {
  CompletionStatus,
  CompletionStep,
  ReduxState,
  Survey as TSurvey,
} from "../../store/types"
import Button from "../Button/Button"
import Styles from "./index.module.scss"

interface SurveyProps {
  survey: TSurvey
  loading?: undefined
}

interface LoadingSurveyProps {
  survey?: undefined
  loading: true
}

const Survey: React.VFC<SurveyProps | LoadingSurveyProps> = ({
  survey,
  loading,
}) => {
  const isSurveyCompleted = useSelector<ReduxState, boolean>(
    (state) =>
      state.profile.profile?.completion[
        CompletionStep.CollegePreferenceSurvey
      ] === CompletionStatus.Completed,
  )

  const { t } = useTranslation()

  return (
    <div className={Styles.container}>
      {!loading && (
        <div className={Styles.iconBox}>
          <Book />
        </div>
      )}
      {loading && <mui.Skeleton width={48} height={48} variant="circular" />}

      <h1 className={Styles.title}>
        {!loading && survey?.title}
        {loading && <mui.Skeleton width={150} className={`mt-2`} />}
      </h1>

      <p className={Styles.description}>
        {!loading && survey?.description}
        {loading &&
          Array.from({ length: 3 }, (_, index) => (
            <mui.Skeleton
              key={index}
              className={`d-flex flex-column my-1`}
              width={175}
              {...(index && { style: { marginTop: `8px` } })}
            />
          ))}
      </p>

      {!loading && (
        <div className={`d-flex justify-content-center`}>
          <Button
            className={Styles.button}
            size={`auto`}
            to={`/surveys/survey/${survey?.id}`}
          >
            {t(
              `surveysScreen.${
                isSurveyCompleted ? `updateAnswers` : `takeSurvey`
              }`,
            )}
          </Button>
        </div>
      )}

      {loading && <mui.Skeleton className={`mt-4`} width={175} height={48} />}
    </div>
  )
}

export default Survey
