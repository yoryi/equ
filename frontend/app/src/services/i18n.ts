import filesize from "filesize"
import i18n from "i18next"
import _ from "lodash"
import { initReactI18next } from "react-i18next"

import { en } from "../translations"

void i18n.use(initReactI18next).init({
  ns: [`common`, `screens`],
  defaultNS: `screens`,
  resources: {
    en,
  },
  lng: `en`,
  fallbackLng: `en`,
  interpolation: {
    format: (value, format, lng) => {
      if (!format) {
        return value
      }

      return format.split(`,`).reduce((value: string, format: string) => {
        switch (format.trim()) {
          case `localeString`:
            return value.toLocaleString()
          case `enIndefiniteArticle`:
          case `enIndefiniteArticleCapitalized`: {
            if (!lng || lng !== `en`) {
              return value
            }

            const letters = [`a`, `e`, `i`, `o`, `u`, `h`]
            const firstLetter = value.substring(0, 1)
            let article = ``
            if (
              letters.find((letter) => firstLetter.toLowerCase() === letter)
            ) {
              article =
                format === `enIndefiniteArticleCapitalized` ? `An` : `an`
            } else {
              article = format === `enIndefiniteArticleCapitalized` ? `A` : `a`
            }

            return `${article} ${value}`
          }
          case `title`: {
            return value
              .toLowerCase()
              .split(` `)
              .map((word: string) => _.upperFirst(word))
              .join(` `)
          }
          case `lowerCase`: {
            return value.toLowerCase()
          }
          case `fileSize`: {
            return filesize(parseInt(value), { round: 0 })
          }
          default:
            return value
        }
      }, value)
    },
  },
})

export default i18n
