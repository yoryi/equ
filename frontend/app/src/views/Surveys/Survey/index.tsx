//Hooks
import * as React from "react"
import { useEffect } from "react"
import { useDispatch,useSelector } from "react-redux"
import { useParams } from "react-router"
//Navigation
import { MemoryRouter, Redirect,Route, Switch } from "react-router"

//Components
import SurveyProgressBar from "../../../components/SurveyProgressBar"
//Actions
import * as actions from "../../../store/actions"
//Types
import { AnswerType,ReduxState, SurveyDetails } from "../../../store/types"
import MultipleChoice from "../MultipleChoice"
//Screens
import SingleChoice from "../SingleChoice"

interface SurveyParams {
  id: string
}

const SurveyView = () => {
  const id = parseInt(useParams<SurveyParams>().id)

  const isTokenLoaded = useSelector<ReduxState, boolean>(
    (state) => state.auth.isTokenLoaded,
  )
  const isSurveyPresent = useSelector<ReduxState, boolean>(
    (state) => !!state.surveys.surveys?.find((survey) => survey.id === id),
  )
  const survey = useSelector<ReduxState, SurveyDetails | null>((state) =>
    state.surveys.survey?.id === id ? state.surveys.survey : null,
  )
  const dispatch = useDispatch()

  useEffect(() => {
    if (!isTokenLoaded || survey) {
      return
    }

    dispatch(actions.getSurvey())
  }, [isTokenLoaded, survey])

  const getScreen = (answerType: AnswerType) => {
    switch (answerType) {
      case AnswerType.SingleChoice:
        return SingleChoice
      case AnswerType.MultipleChoice:
        return MultipleChoice
    }
  }

  if (!survey) {
    return isSurveyPresent ? null : <Redirect to={`/surveys`} />
  }

  return (
    <>
      <MemoryRouter initialEntries={[survey?.questions[0].id.toString()]}>
        <SurveyProgressBar />

        <Switch>
          {survey?.questions.map((question) => (
            <Route
              key={question.id}
              path={question.id.toString()}
              component={getScreen(question.answerType)}
            />
          ))}
        </Switch>
      </MemoryRouter>
    </>
  )
}

export default SurveyView
