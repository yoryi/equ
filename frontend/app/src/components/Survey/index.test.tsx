//Testing
import { render } from "@testing-library/react"
import faker from "faker"
import * as React from "react"
//Translations
import { I18nextProvider } from "react-i18next"
//Redux store
import { Provider } from "react-redux"
//Navigation
import { MemoryRouter } from "react-router"
import configureStore from "redux-mock-store"
//Types
import { MockStore } from "redux-mock-store"

import i18n from "../../services/i18n"
import {
  CompletionStatus,
  CompletionStep,
  Profile,
  ReduxState,
  Role,
  Survey as TSurvey,
} from "../../store/types"
//Components
import Survey from "./"

interface TestStore {
  profile: {
    profile: ReduxState["profile"]["profile"]
  }
}

describe(`Survey component`, () => {
  faker.seed(1000)

  const mockStore = configureStore<TestStore>()

  let store: MockStore<TestStore>, survey: TSurvey
  beforeEach(() => {
    store = mockStore({
      profile: {
        profile: {
          id: faker.random.number(10000),
          cover: null,
          avatar: null,
          privateAvatar: null,
          interests: null,
          firstName: null,
          lastName: null,
          name: null,
          email: null,
          role: Role.Student,
          school: null,
          graduation: null,
          mission: null,
          birthday: null,
          gender: null,
          race: null,
          ethnicity: null,
          hardship: null,
          signUpStep: null,
          completion: {
            [CompletionStep.CollegePreferenceSurvey]: CompletionStatus.Pending,
          },
          isFollowed: false,
          uiSettings: null,
        },
      },
    })

    survey = {
      id: faker.random.number(10000),
      title: faker.lorem.words(2),
      description: faker.lorem.sentence(),
    }
  })

  const wrapper: React.FC = ({ children }) => (
    <I18nextProvider i18n={i18n}>
      <MemoryRouter>
        <Provider store={store}>{children}</Provider>
      </MemoryRouter>
    </I18nextProvider>
  )

  describe(`should render and match the snapshot`, () => {
    describe(`loaded`, () => {
      it(`not completed`, () => {
        const { container } = render(<Survey survey={survey} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })

      it(`completed`, () => {
        store = mockStore({
          profile: {
            profile: {
              ...(store.getState().profile.profile as Profile),
              completion: {
                [CompletionStep.CollegePreferenceSurvey]:
                  CompletionStatus.Completed,
              },
            },
          },
        })

        const { container } = render(<Survey survey={survey} />, {
          wrapper,
        })

        expect(container.firstChild).toMatchSnapshot()
      })
    })

    it(`loading`, () => {
      const { container } = render(<Survey loading />, {
        wrapper,
      })

      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
