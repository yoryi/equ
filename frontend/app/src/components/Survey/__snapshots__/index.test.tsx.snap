// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey component should render and match the snapshot loaded completed 1`] = `
<div
  class="container"
>
  <div
    class="iconBox"
  >
    <svg>
      book.svg
    </svg>
  </div>
  <h1
    class="title"
  >
    at voluptas
  </h1>
  <p
    class="description"
  >
    Quia est quod et reprehenderit veritatis sit.
  </p>
  <div
    class="d-flex justify-content-center"
  >
    <a
      class="primary-button button default-button auto-button blue-button"
      href="/surveys/survey/2331"
    >
      Update Answers
    </a>
  </div>
</div>
`;

exports[`Survey component should render and match the snapshot loaded not completed 1`] = `
<div
  class="container"
>
  <div
    class="iconBox"
  >
    <svg>
      book.svg
    </svg>
  </div>
  <h1
    class="title"
  >
    fugit est
  </h1>
  <p
    class="description"
  >
    Aut atque ea officiis pariatur non perferendis ipsa nisi qui.
  </p>
  <div
    class="d-flex justify-content-center"
  >
    <a
      class="primary-button button default-button auto-button blue-button"
      href="/surveys/survey/2055"
    >
      Take Survey
    </a>
  </div>
</div>
`;

exports[`Survey component should render and match the snapshot loading 1`] = `
<div
  class="container"
>
  <span
    class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-fomgco-MuiSkeleton-root"
    style="width: 48px; height: 48px;"
  />
  <h1
    class="title"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse mt-2 css-1n7gpw1-MuiSkeleton-root"
      style="width: 150px;"
    />
  </h1>
  <p
    class="description"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse d-flex flex-column my-1 css-1n7gpw1-MuiSkeleton-root"
      style="width: 175px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse d-flex flex-column my-1 css-1n7gpw1-MuiSkeleton-root"
      style="width: 175px; margin-top: 8px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse d-flex flex-column my-1 css-1n7gpw1-MuiSkeleton-root"
      style="width: 175px; margin-top: 8px;"
    />
  </p>
  <span
    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse mt-4 css-1n7gpw1-MuiSkeleton-root"
    style="width: 175px; height: 48px;"
  />
</div>
`;
